vue-components/src/components/Dropdown.vue

```diff
@@ -39,28 +39,64 @@
 			@esc="onEsc"
 			ref="menu"
 		/>
+		<ValidationMessage
+			v-if="error"
+			:type="error.type"
+			:message="error.message"
+		/>
 	</div>
 </template>
 
 <script lang="ts">
-import Vue, { PropType, VueConstructor } from 'vue';
+import Vue from 'vue';
+import VueCompositionAPI, { defineComponent, computed, ref } from '@vue/composition-api';
 import { MenuItem } from '@/components/MenuItem';
 import OptionsMenu from '@/components/OptionsMenu.vue';
+import { getFeedbackTypeFromProps, errorProp, ErrorProp } from '@/components/util/ValidationMessageComposition';
 import isEqual from 'lodash.isequal';
+import ValidationMessage from '@/components/ValidationMessage.vue';
+
+Vue.use( VueCompositionAPI );
 
-export default ( Vue as VueConstructor<Vue & { $refs: {
-	menu: InstanceType<typeof OptionsMenu>;
-	select: HTMLElement;
-}; }> ).extend( {
+export default defineComponent( {
 	name: 'Dropdown',
-	data() {
+	setup( props: { error: ErrorProp; menuItems: MenuItem[]; value: MenuItem|null }, { root } ) {
+		const menu = ref<InstanceType<typeof OptionsMenu>|null>( null );
+		const showMenu = ref( false );
+
+		async function startShowingTheMenu(): Promise<void> {
+			showMenu.value = true;
+			await root.$nextTick();
+			menu.value?.resizeMenu();
+		}
+
+		function triggerKeyDown( event: KeyboardEvent ): void {
+			switch ( event.key ) {
+				case 'Enter':
+					if ( !showMenu.value ) {
+						startShowingTheMenu();
+						return;
+					}
+					break;
+				case 'ArrowDown':
+					startShowingTheMenu();
+			}
+			menu.value?.onKeyDown( event );
+		}
```

Comment:
Why was all this code moved into the composition setup?

Reply:
We needed to define the refs in the composition setup to get the TypeScript type working. And I _think_ the problem may afterwards have been how to access them, but I'll check that again.

Reply:
Yes, moving those things back is working now! The diff should now be much smaller :)

Reply:
Or so I thought until I saw the CI complaining. Will have another look this afternoon.

Reply:
Ok, so no. This has to happen inside the `setup`, because only there the ref `menu` has the correct type. It doesn't infer its type correctly when used outside of the setup.

I added a line to the commit message abut that

Reply:
If you want, you can play around with the problem in https://github.com/wmde/wikit/tree/badlyTypedRef

Reply:
I'm not sure if I like this pattern, I will play around with it some more tomorrow to see what can be done to mitigate this... I wonder if Vue 3 can handle this better without relying on pulling everything into setup just because of a ref..

Reply:
Yeah, I suspect this might be a limitation of the plugin. Happy to find a nicer alternative.

OTOH, I could also see us (in the future) transition to the setup-only way of writing single file components:
```
<script setup>
  import { ref } from 'vue'

  export const count = ref(0)
  export const inc = () => count.value++
</script>
```
https://github.com/vuejs/rfcs/blob/sfc-improvements/active-rfcs/0000-sfc-script-setup.md

Reply:
That actually looks beautiful! I'd love it if this rfc is accepted

Reply:
I guess this would be alright for now, but I would open a ticket to investigate this topic further.

Reply:
I can't believe I looked twice into this and didn't figure it out! :facepalm: 

Thanks! Will update the PR